### PR TITLE
gh-111928: make "memo" dict local to scan_once call

### PR DESCRIPTION
**Pull request by aisk**
* Issue: gh-111928

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Thanks for the PR! This looks okay, but I worry there will be a performance regression. Do you know how much of a hit performance takes with this approach?

File: Misc/NEWS.d/next/Library/2023-11-12-21-39-36.gh-issue-111928.SiYBqt.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Make the ``memo`` dict thread safe without the GIL in :mod:`json`.
```
Comment (corona10) [inline Misc/NEWS.d/next/Library/2023-11-12-21-39-36.gh-issue-111928.SiYBqt.rst]: NEWS.d is not needed since there is no user side changes.

**Review by corona10 — COMMENTED** [review]
~Would you like to add @colesbury as co-author since the PR is cherry-picked from his own repo?~

Oh you did already.

**Review by colesbury — APPROVED** [review]
This lgtm.

@Fidget-Spinner, performance impact is <1% for parsing tiny strings of JSON (and less for larger JSON strings).

**Comment by colesbury** [comment]
Meant to link to a quick and dirty JSON benchmark https://gist.github.com/colesbury/dd888920a1eacb97794fb6be2446dc10:

**Review by corona10 — APPROVED** [review]

**Pull request merged**